contracts/remarkables/src/contract.rs

```diff
@@ -0,0 +1,699 @@
+            },
+            info.funds,
+            "remarkables_cw721".to_string(),
+        )?,
+        INSTANTIATE_CW721_REPLY_ID,
+    );
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_INSTANTIATE)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_ADMIN, msg.admin)
+        .add_attribute(ATTRIBUTE_CW721_CODE_ID, msg.cw721_code_id)
+        .add_submessage(cw721_submessage))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut<DesmosQuery>,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    msg.validate()?;
+    match msg {
+        ExecuteMsg::MintTo {
+            post_id,
+            remarkables_uri,
+            rarity_level,
+        } => execute_mint_to(deps, info, rarity_level, post_id.into(), remarkables_uri),
+        ExecuteMsg::UpdateAdmin { new_admin } => execute_update_admin(deps, info, new_admin),
+        ExecuteMsg::UpdateRarityMintFee {
+            rarity_level,
+            new_fees,
+        } => execute_update_rarity_mint_fees(deps, info, rarity_level, new_fees),
+    }
+}
+
+fn execute_mint_to(
+    deps: DepsMut<DesmosQuery>,
+    info: MessageInfo,
+    rarity_level: u32,
+    post_id: u64,
+    remarkables_uri: String,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    let rarity = RARITY.load(deps.storage, rarity_level)?;
+    // Check if rarity mint fees is enough
+    if !is_enough_fees(info.funds, rarity.mint_fees) {
+        return Err(ContractError::MintFeesNotEnough {});
+    }
+    // Check if post reaches the eligible threshold
+    check_eligibility(
+        deps.storage,
+        deps.querier.deref(),
+        info.sender.clone(),
+        post_id,
+        rarity.engagement_threshold,
+    )?;
+    // Create the cw721 message to send to mint the remarkables
+    let mint_msg = Cw721ExecuteMsg::<Empty, Empty>::Mint(MintMsg::<Empty> {
+        token_id: post_id.to_string(),
+        owner: info.sender.clone().into(),
+        token_uri: Some(remarkables_uri.clone()),
+        extension: Empty {},
+    });
+    let wasm_execute_mint_msg = wasm_execute(CW721_ADDRESS.load(deps.storage)?, &mint_msg, vec![])?;
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_MINT_TO)
+        .add_attribute(ATTRIBUTE_SENDER, &info.sender)
+        .add_attribute(ATTRIBUTE_RARITY_LEVEL, rarity_level.to_string())
+        .add_attribute(ATTRIBUTE_TOKEN_ID, post_id.to_string())
+        .add_attribute(ATTRIBUTE_RECIPIENT, &info.sender)
+        .add_attribute(ATTRIBUTE_TOKEN_URI, remarkables_uri)
+        .add_message(wasm_execute_mint_msg))
+}
+
+fn is_enough_fees(funds: Vec<Coin>, requireds: Vec<Coin>) -> bool {
+    if requireds.len() == 0 {
+        return true;
+    }
+    // It takes O(n^2) time complexity but both list are extremely small
+    for required in requireds.iter() {
+        if has_coins(&funds, &required) {
```

Comment:
I think that this if should be `!has_coins` and should return `false` and return `true` at line 178 otherwise if we have two lists of coins like this:
1. founds: [100udsm]
2. requireds: [100uds, 100uatom]  

This function will wrongly return true.
Please also add a test with such case to have it covered.

Reply:
Thanks for spotting the issue, I thought required fees should be `or` condition.

Reply:
Thanks for spotting the issue. I thought required fees should be `or` condition, instead of `and`. I would like to confirm which condition in mint fees should be.
cc @bragaz

Reply:
Fixed